runtime/reconcilers/project_parser.go

```diff
@@ -293,14 +293,16 @@ func (r *ProjectParserReconciler) reconcileParser(ctx context.Context, inst *dri
 		return parseErrsErr
 	}
 
+	// not setting restartController=true when diff is actually nil prevents infinite restarts
+	restartController := diff != nil && (diff.ModifiedDotEnv || diff.Reloaded)
 	// Treat reloads the same as a fresh parse (where there's no diff)
 	if diff != nil && diff.Reloaded {
 		diff = nil
 	}
 
 	// Update state from rill.yaml and .env
 	if diff == nil || diff.ModifiedDotEnv {
-		err := r.reconcileProjectConfig(ctx, parser)
+		err := r.reconcileProjectConfig(ctx, parser, restartController)
 		if err != nil {
 			return err
 		}
```

Comment:
If setting `restartController` to `true`, this reconciler's `ctx` will be cancelled right after, so I guess it should return here? And since you also restart when `diff.Reloaded`, I guess the handling for when `diff == nil` later in this file is redundant?